File: src/main/java/com/yupi/airouter/controller/PluginController.java
/**
 * 插件管理控制器
 *
 * @author <a href="https://codefather.cn">编程导航学习圈</a>
 */
package com.yupi.airouter.controller;

import com.yupi.airouter.annotation.AuthCheck;
import com.yupi.airouter.common.BaseResponse;
import com.yupi.airouter.common.ResultUtils;
import com.yupi.airouter.constant.UserConstant;
import com.yupi.airouter.exception.BusinessException;
import com.yupi.airouter.exception.ErrorCode;
import com.yupi.airouter.model.dto.plugin.PluginExecuteRequest;
import com.yupi.airouter.model.dto.plugin.PluginUpdateRequest;
import com.yupi.airouter.model.entity.User;
import com.yupi.airouter.model.vo.PluginConfigVO;
import com.yupi.airouter.model.vo.PluginExecuteVO;
import com.yupi.airouter.service.PluginService;
import com.yupi.airouter.service.UserService;
import io.swagger.v3.oas.annotations.Operation;
import jakarta.annotation.Resource;
import jakarta.servlet.http.HttpServletRequest;
import lombok.extern.slf4j.Slf4j;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@Slf4j
@RestController
@RequestMapping("/plugin")
public class PluginController {

    @Resource
    private PluginService pluginService;

    @Resource
    private UserService userService;

    /**
     * 获取所有插件列表（管理员）
     */
    @GetMapping("/list")
    @Operation(summary = "获取所有插件列表")
    @AuthCheck(mustRole = UserConstant.ADMIN_ROLE)
    public BaseResponse<List<PluginConfigVO>> listPlugins() {
        List<PluginConfigVO> pluginList = pluginService.listAllPlugins();
        return ResultUtils.success(pluginList);
    }

    /**
     * 获取启用的插件列表
     */
    @GetMapping("/list/enabled")
    @Operation(summary = "获取启用的插件列表")
    public BaseResponse<List<PluginConfigVO>> listEnabledPlugins() {
        List<PluginConfigVO> pluginList = pluginService.listEnabledPlugins();
        return ResultUtils.success(pluginList);
    }

    /**
     * 获取插件详情
     */
    @GetMapping("/get")
    @Operation(summary = "获取插件详情")
    public BaseResponse<PluginConfigVO> getPlugin(@RequestParam String pluginKey) {
        if (pluginKey == null || pluginKey.isEmpty()) {
            throw new BusinessException(ErrorCode.PARAMS_ERROR, "插件标识不能为空");
        }
        PluginConfigVO pluginConfig = pluginService.getPluginByKey(pluginKey);
        if (pluginConfig == null) {
            throw new BusinessException(ErrorCode.NOT_FOUND_ERROR, "插件不存在");
        }
        return ResultUtils.success(pluginConfig);
    }

    /**
     * 更新插件配置（管理员）
     */
    @PostMapping("/update")
    @Operation(summary = "更新插件配置")
    @AuthCheck(mustRole = UserConstant.ADMIN_ROLE)
    public BaseResponse<Boolean> updatePlugin(@RequestBody PluginUpdateRequest request) {
        if (request == null) {
            throw new BusinessException(ErrorCode.PARAMS_ERROR);
        }
        boolean result = pluginService.updatePlugin(request);
        return ResultUtils.success(result);
    }

    /**
     * 启用插件（管理员）
     */
    @PostMapping("/enable")
    @Operation(summary = "启用插件")
    @AuthCheck(mustRole = UserConstant.ADMIN_ROLE)
    public BaseResponse<Boolean> enablePlugin(@RequestParam String pluginKey) {
        if (pluginKey == null || pluginKey.isEmpty()) {
            throw new BusinessException(ErrorCode.PARAMS_ERROR, "插件标识不能为空");
        }
        boolean result = pluginService.enablePlugin(pluginKey);
        return ResultUtils.success(result);
    }

    /**
     * 禁用插件（管理员）
     */
    @PostMapping("/disable")
    @Operation(summary = "禁用插件")
    @AuthCheck(mustRole = UserConstant.ADMIN_ROLE)
    public BaseResponse<Boolean> disablePlugin(@RequestParam String pluginKey) {
        if (pluginKey == null || pluginKey.isEmpty()) {
            throw new BusinessException(ErrorCode.PARAMS_ERROR, "插件标识不能为空");
        }
        boolean result = pluginService.disablePlugin(pluginKey);
        return ResultUtils.success(result);
    }

    /**
     * 执行插件
     */
    @PostMapping("/execute")
    @Operation(summary = "执行插件")
    public BaseResponse<PluginExecuteVO> executePlugin(@RequestBody PluginExecuteRequest request,
                                                       HttpServletRequest httpRequest) {
        if (request == null) {
            throw new BusinessException(ErrorCode.PARAMS_ERROR);
        }

        // 获取当前用户
        User loginUser = userService.getLoginUser(httpRequest);
        Long userId = loginUser != null ? loginUser.getId() : null;

        PluginExecuteVO result = pluginService.executePlugin(request, userId);
        return ResultUtils.success(result);
    }

    /**
     * 重新加载插件（管理员）
     */
    @PostMapping("/reload")
    @Operation(summary = "重新加载插件")
    @AuthCheck(mustRole = UserConstant.ADMIN_ROLE)
    public BaseResponse<Boolean> reloadPlugin(@RequestParam String pluginKey) {
        if (pluginKey == null || pluginKey.isEmpty()) {
            throw new BusinessException(ErrorCode.PARAMS_ERROR, "插件标识不能为空");
        }
        pluginService.reloadPlugin(pluginKey);
        return ResultUtils.success(true);
    }

    /**
     * 重新加载所有插件（管理员）
     */
    @PostMapping("/reload/all")
    @Operation(summary = "重新加载所有插件")
    @AuthCheck(mustRole = UserConstant.ADMIN_ROLE)
    public BaseResponse<Boolean> reloadAllPlugins() {
        pluginService.initPlugins();
        return ResultUtils.success(true);
    }
}

File: go-backend/internal/controller/plugin_controller.go
package controller

import (
	"log"
	"strings"

	"github.com/gin-gonic/gin"

	"github.com/yupi/airouter/go-backend/internal/common"
	"github.com/yupi/airouter/go-backend/internal/errno"
	"github.com/yupi/airouter/go-backend/internal/model/dto"
	"github.com/yupi/airouter/go-backend/internal/service"
)

type PluginController struct {
	pluginService *service.PluginService
	userService   *service.UserService
}

func NewPluginController(pluginService *service.PluginService, userService *service.UserService) *PluginController {
	return &PluginController{
		pluginService: pluginService,
		userService:   userService,
	}
}

func (p *PluginController) ListPlugins(c *gin.Context) {
	result, err := p.pluginService.ListAllPlugins()
	if err != nil {
		p.handleError(c, err)
		return
	}
	common.Success(c, result)
}

func (p *PluginController) ListEnabledPlugins(c *gin.Context) {
	result, err := p.pluginService.ListEnabledPlugins()
	if err != nil {
		p.handleError(c, err)
		return
	}
	common.Success(c, result)
}

func (p *PluginController) GetPlugin(c *gin.Context) {
	pluginKey := strings.TrimSpace(c.Query("pluginKey"))
	if pluginKey == "" {
		common.Error(c, errno.ParamsError.Code, "插件标识不能为空")
		return
	}
	result, err := p.pluginService.GetPluginByKey(pluginKey)
	if err != nil {
		p.handleError(c, err)
		return
	}
	common.Success(c, result)
}

func (p *PluginController) UpdatePlugin(c *gin.Context) {
	var request dto.PluginUpdateRequest
	if err := c.ShouldBindJSON(&request); err != nil {
		log.Printf("update plugin bind failed: %v", err)
		common.Error(c, errno.ParamsError.Code, errno.ParamsError.Message)
		return
	}
	result, err := p.pluginService.UpdatePlugin(request)
	if err != nil {
		p.handleError(c, err)
		return
	}
	common.Success(c, result)
}

func (p *PluginController) EnablePlugin(c *gin.Context) {
	pluginKey := strings.TrimSpace(c.Query("pluginKey"))
	if pluginKey == "" {
		common.Error(c, errno.ParamsError.Code, "插件标识不能为空")
		return
	}
	result, err := p.pluginService.EnablePlugin(pluginKey)
	if err != nil {
		p.handleError(c, err)
		return
	}
	common.Success(c, result)
}

func (p *PluginController) DisablePlugin(c *gin.Context) {
	pluginKey := strings.TrimSpace(c.Query("pluginKey"))
	if pluginKey == "" {
		common.Error(c, errno.ParamsError.Code, "插件标识不能为空")
		return
	}
	result, err := p.pluginService.DisablePlugin(pluginKey)
	if err != nil {
		p.handleError(c, err)
		return
	}
	common.Success(c, result)
}

func (p *PluginController) ExecutePlugin(c *gin.Context) {
	var request dto.PluginExecuteRequest
	if err := c.ShouldBindJSON(&request); err != nil {
		log.Printf("execute plugin bind failed: %v", err)
		common.Error(c, errno.ParamsError.Code, errno.ParamsError.Message)
		return
	}
	var userID *int64
	if loginUser, err := p.userService.GetLoginUser(c); err == nil && loginUser != nil && loginUser.ID > 0 {
		userID = &loginUser.ID
	}
	result, err := p.pluginService.ExecutePlugin(request, userID)
	if err != nil {
		p.handleError(c, err)
		return
	}
	common.Success(c, result)
}

func (p *PluginController) ReloadPlugin(c *gin.Context) {
	pluginKey := strings.TrimSpace(c.Query("pluginKey"))
	if pluginKey == "" {
		common.Error(c, errno.ParamsError.Code, "插件标识不能为空")
		return
	}
	if err := p.pluginService.ReloadPlugin(pluginKey); err != nil {
		p.handleError(c, err)
		return
	}
	common.Success(c, true)
}

func (p *PluginController) ReloadAllPlugins(c *gin.Context) {
	if err := p.pluginService.InitPlugins(); err != nil {
		p.handleError(c, err)
		return
	}
	common.Success(c, true)
}

func (p *PluginController) handleError(c *gin.Context, err error) {
	if bizErr, ok := errno.AsBusinessError(err); ok {
		common.Error(c, bizErr.Code, bizErr.Message)
		return
	}
	common.Error(c, errno.SystemError.Code, errno.SystemError.Message)
}
